src/distribution/cauchy.rs

```diff
@@ -0,0 +1,238 @@
+        let p = vec![
+            0.03488327519822364,
+            0.03744822190397538,
+            0.03843742021842001,
+            0.039176601376466544,
+            0.03963391578942141,
+            0.039788735772973836,
+            0.03963391578942141,
+            0.039176601376466544,
+            0.03744822190397538,
+            0.03183098861837907,
+            0.015527311521160521,
+        ];
+
+        assert::close(
+            &x.iter().map(|&x| d.density(x)).collect::<Vec<_>>(),
+            &p,
+            1e-15,
+        );
+    }
+
+    #[test]
+    fn distribution() {
+        let d = new!(2.0, 8.0);
+        let x = vec![
+            -1.0, 0.0, 0.01, 0.05, 0.1, 0.15, 0.25, 0.5, 1.0, 1.5, 2.0, 3.0, 4.0,
+        ];
+        let p = vec![
+            0.3857997487800918,
+            0.4220208696226307,
+            0.4223954618429798,
+            0.4238960166273086,
+            0.4257765641957529,
+            0.42766240385764065,
+            0.43144951512041,
+            0.44100191513247144,
+            0.46041657583943446,
+            0.48013147569445913,
+            0.5,
+            0.5395834241605656,
+            0.5779791303773694,
+        ];
+
+        assert::close(
+            &x.iter().map(|&x| d.distribution(x)).collect::<Vec<_>>(),
+            &p,
+            1e-15,
+        );
+    }
+
+    #[test]
+    fn entropy() {
+        use std::f64::consts::PI;
+        assert_eq!(new!(2.0, 1.0).entropy(), (PI * 4.0).ln());
+        assert::close(new!(3.0, 5.2).entropy(), 4.1796828725566719243, 1e-15);
+    }
+
+    #[test]
+    fn inverse() {
+        let d = new!(2.0, 3.0);
+        let x = vec![
+            -7.2330506115257585,
+            -0.9999999999999996,
+            2.0,
+            5.0,
+            11.233050611525758,
+        ];
+        let p = vec![0.1, 0.25, 0.5, 0.75, 0.9];
+
+        assert::close(
+            &p.iter().map(|&p| d.inverse(p)).collect::<Vec<_>>(),
+            &x,
+            1e-14,
+        );
+
+        assert!(d.inverse(0.0) < -1e16);
+        assert!(d.inverse(1.0) > 1e16);
+    }
+
+    #[test]
+    fn median() {
+        assert_eq!(new!(2.0, 1.0).median(), 2.0);
+    }
+
+    #[test]
+    fn modes() {
+        assert_eq!(new!(2.0, 1.0).modes(), vec![2.0]);
+    }
+
+    #[test]
+    fn sampling() {
+        let n = 100000;
+        let d = Cauchy::new(35.4, 12.3);
+        let mut source = source::Xorshift128Plus::new([42, 69]);
+
+        // Estimate the Kullback-Leibler divergence `KL(samples || d)` based on
+        // `n` samples and assert that the estimate is close to zero.
+        // This test is reproducible because we use a fixed random seed.
+        let cross_entropy_estimate = -(0..n)
+            .map(|_| d.density(d.sample(&mut source)).ln())
+            .sum::<f64>()
+            / n as f64;
+        let kl_divergence_estimate = cross_entropy_estimate - d.entropy();
+        assert!(kl_divergence_estimate.abs() < 0.01);
```

Comment:
Not sure if 0.01 qualifies as “close to zero.”

Reply:
It's about two standard deviations, so I think, if anything, the threshold is a bit ambitious and could be increased further (i.e., if this test fails for a different random seed or a different sampling algorithm, then that doesn't necessarily mean that the sampling algorithm is wrong).

In detail: We're estimating the quantity `D := H(q, p) - H(p)` where `q` is the sampling distribution and `p` is the distribution that's used for calculating the density. If `p = q` (which we want to test here), then `D = 0` exactly. However, we're only *estimating* `D` based on a finite number `n` of samples, i.e., we're really sampling a random variable `D_n` whose mean is zero but whose standard deviation works out to be `PI / sqrt(3 * n)`. We use `n = 100000` here, so the standard deviation is about `0.0057`.